components/remote_settings/src/client.rs

```diff
@@ -18,6 +19,205 @@ const HEADER_BACKOFF: &str = "Backoff";
 const HEADER_ETAG: &str = "ETag";
 const HEADER_RETRY_AFTER: &str = "Retry-After";
 
+/// Internal Remote settings client API
+pub struct RemoteSettingsClient {
+    // This is immutable, so it can be outside the mutex
+    collection_name: String,
+    inner: Mutex<RemoteSettingsClientInner>,
+}
+
+struct RemoteSettingsClientInner {
+    storage: Storage,
+    api_client: ApiClient,
+}
+
+impl RemoteSettingsClient {
+    pub fn new(
+        server_url: Url,
+        bucket_name: String,
+        collection_name: String,
+        storage: Storage,
+    ) -> Result<Self> {
+        let api_client = ApiClient::new(server_url, &bucket_name, &collection_name)?;
+        Ok(Self {
+            collection_name,
+            inner: Mutex::new(RemoteSettingsClientInner {
+                storage,
+                api_client,
+            }),
+        })
+    }
+
+    pub fn collection_name(&self) -> &str {
+        &self.collection_name
+    }
+
+    pub fn get_records(&self, sync_if_empty: bool) -> Result<Option<Vec<RemoteSettingsRecord>>> {
+        let mut inner = self.inner.lock();
+        let base_url = inner.api_client.base_url();
+
+        let cached_records = inner.storage.get_records(&base_url)?;
+        if cached_records.is_some() || !sync_if_empty {
+            return Ok(cached_records);
+        }
+
+        let records = inner.api_client.get_records(None)?;
+        inner.storage.set_records(&base_url, &records)?;
+        Ok(Some(records))
+    }
+
+    pub fn sync(&self) -> Result<()> {
+        let mut inner = self.inner.lock();
+        let base_url = inner.api_client.base_url();
+        let mtime = inner.storage.get_last_modified_timestamp(&base_url)?;
+        let records = inner.api_client.get_records(mtime)?;
+        inner.storage.set_records(&base_url, &records)
+    }
+
+    /// Downloads an attachment from [attachment_location]. NOTE: there are no guarantees about a
+    /// maximum size, so use care when fetching potentially large attachments.
+    pub fn get_attachment(&self, attachment_location: &str) -> Result<Vec<u8>> {
+        self.inner
+            .lock()
+            .api_client
+            .get_attachment(attachment_location)
+    }
+
+    pub fn update_config(&self, server_url: Url, bucket_name: String) -> Result<()> {
+        let mut inner = self.inner.lock();
+        inner.api_client = ApiClient::new(server_url, &bucket_name, &self.collection_name)?;
+        inner.storage.empty()
+    }
+}
+
+/// Client for Remote settings API requests
+struct ApiClient {
+    /// Base URL for API requests
+    ///
+    /// This is something like
+    /// `https://[server-url]/v1/buckets/[bucket-name]/collections/[collection-name]/"
+    base_url: Url,
+    remote_state: RemoteState,
+}
+
+impl ApiClient {
+    fn new(server_url: Url, bucket_name: &str, collection_name: &str) -> Result<Self> {
+        let base_url = server_url.join(&format!(
+            "v1/buckets/{bucket_name}/collections/{collection_name}/"
```

Comment:
I wanted something that would change when the bucket changed, so that we could use this URL to test if the cached data was still valid.  I should probably use a different name for this though, maybe `collection_url`?  What's your feeling on this one?

Reply:
I went ahead and renamed this to `collection_url`.  Hopefully that makes sense, but I'm open to just storing the `base_url` to match how other implementations do it.

Reply:
`collection_url` works.
However, please put the `v1` as part of the server URL.
(eg. https://github.com/mozilla-extensions/remote-settings-devtools/blob/15245fdb8e4fb5875a7866c7328ded6671ab466c/extension/experiments/remotesettings/api.js#L8-L11)

Reply:
Sounds good, but let me do this in a followup.  It's a small change, but not trivial since the URL is part of the public API.